ProcessLib/ComponentTransport/ComponentTransportFEM.h

```diff
@@ -358,27 +373,328 @@ class LocalAssemblerData : public ComponentTransportLocalAssemblerInterface
+                    MaterialPropertyLib::molecular_diffusion);
+
+            auto const& K =
+                _process_data.porous_media_properties.getIntrinsicPermeability(
+                    t, pos).getValue(t, pos, 0.0, 0.0);
+            // Use the viscosity model to compute the viscosity
+            auto const mu = _process_data.fluid_properties->getValue(
+                MaterialLib::Fluid::FluidPropertyType::Viscosity, vars);
+
+            GlobalDimMatrixType const K_over_mu = K / mu;
+            GlobalDimVectorType const velocity =
+                _process_data.has_gravity
+                    ? GlobalDimVectorType(-K_over_mu *
+                                          (dNdx * local_p - density * b))
+                    : GlobalDimVectorType(-K_over_mu * dNdx * local_p);
+
+            const double drho_dp = _process_data.fluid_properties->getdValue(
+                MaterialLib::Fluid::FluidPropertyType::Density,
+                vars,
+                MaterialLib::Fluid::PropertyVariableType::p);
+
+            const double drho_dC = _process_data.fluid_properties->getdValue(
+                MaterialLib::Fluid::FluidPropertyType::Density,
+                vars,
+                MaterialLib::Fluid::PropertyVariableType::C);
+            double const velocity_magnitude = velocity.norm();
+            GlobalDimMatrixType const hydrodynamic_dispersion =
+                velocity_magnitude != 0.0
+                    ? GlobalDimMatrixType(
+                          (porosity * molecular_diffusion_coefficient +
+                           solute_dispersivity_transverse *
+                               velocity_magnitude) *
+                              I +
+                          (solute_dispersivity_longitudinal -
+                           solute_dispersivity_transverse) /
+                              velocity_magnitude * velocity *
+                              velocity.transpose())
+                    : GlobalDimMatrixType(
+                          (porosity * molecular_diffusion_coefficient +
+                           solute_dispersivity_transverse *
+                               velocity_magnitude) *
+                          I);
+
+            // matrix assembly
+            local_M.noalias() += w * N.transpose() * retardation_factor *
+                                     porosity * density * N +
+                                 w * N.transpose() * retardation_factor *
+                                     porosity * C_int_pt * drho_dC * N;
+
+            // coupling term
+            {
+                double p0_int_pt = 0.0;
+                NumLib::shapeFunctionInterpolate(local_p0, N, p0_int_pt);
+
+                local_K.noalias() +=
+                    (-dNdx.transpose() * velocity * density * N +
+                     dNdx.transpose() * density * hydrodynamic_dispersion *
+                         dNdx +
+                     N.transpose() * decay_rate * porosity *
```

Comment:
You can put these two calculations together as
```
                     N.transpose() * ( decay_rate * porosity *
                         retardation_factor * density +
                        porosity * retardation_factor * drho_dp *
                         (p_int_pt - p0_int_pt) * w / dt ) *N) 
```
to save matrix computing operation.

Reply:
Changed.